File: 0582-kill-process/0582-kill-process.java
class Solution {

    boolean[] visited = new boolean[50001]; 
    HashMap<Integer,Integer> hashMap = new HashMap<>(); 

    public void dfs(List<List<Integer>> adj, int kill, List<Integer> answer){
        answer.add(kill); 
        for(Integer next : adj.get(kill)){
            if(!visited[next] && hashMap.get(kill) != next){
                visited[next] = true;
                dfs(adj,next,answer); 
            }
        }
        
    }
    public List<Integer> killProcess(List<Integer> pid, List<Integer> ppid, int kill) {
        List<Integer> answer = new ArrayList<>(); 
        List<List<Integer>> adj = new ArrayList<>(); 
        for(int i = 0; i < 50001; i++){
            adj.add(new ArrayList<>()); 
        }

        for(int i = 0; i < pid.size(); i++){
            if(ppid.get(i) == 0) continue; 

            adj.get(pid.get(i)).add(ppid.get(i)); 
            adj.get(ppid.get(i)).add(pid.get(i)); 
            hashMap.put(pid.get(i), ppid.get(i)); 
        }
        visited[kill] = true; 
        dfs(adj,kill,answer); 
        return answer; 
    }
}

File: 0582-kill-process/0582-kill-process.cpp
class Solution {
public:
    map<int,int> hashMap; 
    vector<int> answer; 
    void dfs(vector<vector<int>>& adj, int hashNode){
        answer.push_back(hashNode); 
        for(int next : adj[hashNode]){
            dfs(adj,next); 
        }
    }
    vector<int> killProcess(vector<int>& pid, vector<int>& ppid, int kill) {
        
        
        for(int i = 0; i < pid.size(); i++){
            hashMap[pid[i]] = i; 
        }

        vector<vector<int>> adj(pid.size()+1); 
        hashMap[0] = pid.size(); 
        for(int i = 0; i < pid.size(); i++){
            //if(ppid[i] == 0) continue; 
            int to = hashMap[pid[i]];
            int from = hashMap[ppid[i]]; 
            adj[from].push_back(to); 
        }
        
        dfs(adj,hashMap[kill]); 

        for(int i = 0; i < answer.size(); i++){
            answer[i] = pid[answer[i]]; 
        }

        return answer; 
    }
};